common/djangoapps/track/eventshim.py

```diff
@@ -0,0 +1,380 @@
+"""
+EventShims that create
+"""
+
+# pylint: disable=missing-docstring
+
+import json
+import logging
+
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import UsageKey
+
+log = logging.getLogger(__name__)
+
+
+class DottedPathMapping(object):
+    """
+    Dictionary-like object for creating keys of dotted paths.
+
+    If a key is created that ends with a dot, it will be treated as a path
+    prefix.  Any value whose prefix matches the dotted path can be used
+    as a key for that value, but only the most specific match will
+    be used.
+    """
+
+    def __init__(self, registry=None):
+        self._match_registry = {}
+        self._prefix_registry = {}
```

Comment:
I found it odd that the existence of a `.` indicated whether it was a prefix match or an exact match.  Since this is non-standard, I see that you have added ample comments regarding this.  This is why I thought a regex-based dictionary would be more readily understandable as something a client would expect.
However, your concern regarding performance is definitely valid.  Especially if we imagine more transformers to be added over time.

1. Agreed.
2. The registry would then be a list instead of a dictionary, allowing us to configure the order - similar to URLConfs.
3. May not be required at this time since I see only 1 prefix transformer registered today.  A comment would be good though.
4. The regex-based registry could still be as simple as accepting a well-constrained regex string as the key and having the registry call `re.match`.